[BOLT][AArch64] Fixes assertion errors occurred when perf2bolt was executed

kaadam opened the pull request:
BOLT only checks for the most common indirect branch pattern during the branch analyzation.
Extended the logic with two other indirect patterns which slightly differ from the expected one.
Those patterns may be hit when statically linking libc (pattern 2 requires 'lld' linker).

As a workaround mark them as UNKNOWN branch for now. 

Fixes: #83114

kaadam committed:
[BOLT][AArch64] Fixes assertion errors occurred when perf2bolt was executed

BOLT only checks for the most common indirect branch pattern during
the branch analyzation. Needs to extend the current logic with other cases
which slightly differs from the expected one.
This instruction sequences comes from libc,
it occurs when the binary is static.

As a workaround mark it as UNKNOWN branch and add support for them
in a follow up PR later.

Fixes: #83114

yota9 commented:
Hello, thank you for a change.
But aren't these changes related to JT? We're loading table address, offset, add them and jump - I'm not completely sure why we want to skip it.
Also please add proper test.

kaadam commented:
Hi @yota9, Sorry for the delayed response.
Started to work on to handle these two patterns maybe you can give me some pointers about them.  
Preprocessing indirect branch we should extract PCRelBase, and JumpTable (label address) + Offset if it is possible.

Considering this pattern, this adr  instruction its address could be PCRelbase, as I tried to follow the original logic maybe it could be "JT" as well. But we only have the address of the entry in sigall_set "section", should we know about the label address of the sigall_set?

```
      //   adr     x6, 0x219fb0 <sigall_set+0x88> => PCRelBase, JT?
      //   add     x6, x6, x14, lsl #2
      //   ldr     w7, [x6]
      //   add     x6, x6, w7, sxtw
      //   br      x6
```
a snippet of the print-cfg:

```
.Ltmp7286 (10 instructions, align : 1)
  CFI State : 0
  Predecessors: .LFT5416
    00000268:   ldp     q5, q6, [x1], #0x20
    0000026c:   stp     q1, q2, [x3], #0x20
    00000270:   and     x3, x3, #0xfffffffffffffff0
    00000274:   sub     x2, x2, #0x20
    00000278:   hint    #0x0 # NOP: 1
    0000027c:   adr     x6, #-365276
    00000280:   add     x6, x6, x14, lsl #2
    00000284:   ldr     w7, [x6]
    00000288:   add     x6, x6, w7, sxtw
    0000028c:   br      x6 # UNKNOWN CONTROL FLOW
  CFI State: 0
```

Binary attached: 
[bubblesort_arm64_static.zip](https://github.com/llvm/llvm-project/files/14758180/bubblesort_arm64_static.zip)

yota9 commented:
Hi @kaadam ! The signal_set is just closest (from above) symbol in the binary, probably it has nothing to do with JT. Consider 0x219fb0 as a base address.

paschalis-mpeis commented:
Two cases are identified in this patch, and TMU the first of which (`(ShiftVal != 2)`)  is hit by users more often. Virtually any stage-0 binary that is statically linked should hit this.

With the patch as is (ie returning 'false'), BOLT won't be able to build the CFG for such functions, meaning less optimizations but a stable binary regardless.

@kaadam has made some good progress locally, to properly parse these patterns, but it's still incomplete.

So I would suggest that we do this on 2 steps:

#### 1) have a quick workaround in place:
Merge for now something like the latest commit, plus some **added test**:

Any .c file, statically linked with libc triggers the first error.
Maybe we can emit some additional warning before returning false, or use `--print-cfg` or `-v=1` and FileCheck that. 

test.c:
```c
int main() { return 1; }
```

in lit test:
```bash
clang -O0 test.c -static -o out
llvm-bolt out -o out.bolt
```

#### 2) use follow-up patches for a proper solution:
Parse the 2 patterns, extract any relevant info (JumpTable/Offset/Scale/Base), and adjust tests as needed.
The `DefJTBaseAdd` being an `ADR` would need an extra test. There's already good local progress on this direction..

paschalis-mpeis commented on bolt/test/AArch64/test-indirect-branch.c:
```diff
@@ -0,0 +1,33 @@
+
+
+// REQUIRES: system-linux
+// RUN: %clang %s -o %t.exe -Wl,-q -static -fuse-ld=lld --target=aarch64-linux
```
Re `--target=aarch64-linux`, there was a recent change that makes this obsolete. So if merged like that, the tests would fail.

Can you please:
1. rebase to the latest main, which includes a handful of commits that did that change (details if you want here [comment here](https://github.com/llvm/llvm-project/pull/93584#issuecomment-2147063138)),
2. and then replace that with `--target=aarch64-unknown-linux-gnu` ?

kaadam replied:
Good to know. Thanks for you comment. Updated!

paschalis-mpeis commented on bolt/test/AArch64/test-indirect-branch.c:
```diff
@@ -0,0 +1,33 @@
+// REQUIRES: system-linux
+// RUN: %clang %s -o %t.exe -Wl,-q -static -fuse-ld=lld --target=aarch64-linux
+// RUN: llvm-bolt %t.exe -o %t.bolt --print-cfg \
+// RUN:  --debug 2>&1 | FileCheck --match-full-lines %s
```
It looks like the `--debug` flag is making tests very slow?

And IMHU we could 'relax' the assertion a bit with a BOLT-WARNING (using just `err()` as in other places in BOLT), which would allow you removing the `--debug` flag from the test, making it faster.

kaadam replied:
Yes, probably it's better to use errs() indeed, we have less output.

paschalis-mpeis commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -752,6 +765,20 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+      //  adr     x12, 0x247b30 <__gettextparse+0x5b0>
+      //  add     x13, x12, w13, sxth #2
+      //  br      x13
+      LLVM_DEBUG(
```
Maybe we could use just `err()` for a BOLT-WARNING? (same applies to previous pattern. see comment on tests).

kaadam committed:
Enable this test only on AArch64 target

kaadam commented:
Seems CIs are happy for now. @yota9 could you take a look at it again please? What do you think about it? Thanks!

yota9 commented:
Hi! I don't really mind to add stubs like this, although I would prefer asm test, since it is too deeply relies on the compiler code generation. Let also @rafaelauler tell his opinion on this :)

paschalis-mpeis commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -706,8 +706,20 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
-    if (ShiftVal != 2)
-      llvm_unreachable("Failed to match indirect branch! (fragment 2)");
+    if (ShiftVal != 2) {
+      // TODO: handle that case where ShiftVal != 2.
```
nit: 'handle the pattern where..'

kaadam replied:
Thanks for the review. Done.

paschalis-mpeis commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -752,6 +764,19 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    if (DefJTBaseAdd->getOpcode() == AArch64::ADR) {
+      // TODO: handle that case when we do not have adrp/add pair.
+      // It also occurs when the binary is static.
+      //  adr     x13, 0x215a18 <_nl_value_type_LC_COLLATE+0x50>
```
nit: '..handle the pattern where there is no adrp...'

kaadam replied:
Thanks for the review. Done.

paschalis-mpeis reviewed (commented):
#### Regarding tests:
> I would prefer asm test, since it is too deeply relies on the compiler code generation.

I don't have a strong opinion, but I do agree that an asm specific test would be a better test.


#### Regarding the safety of relaxation of assertions to BOLT-WARNINGS:
With some small local tests, I've seen that BOLT simply won't modify such functions, so we won't have any correctness issues (hence the initial suggestion). But I wanted to be sure we weren't missing any corner cases, so I've discussed this yesterday during BOLT 'Office Hours' with Rafael and Maks.

From the discussion, such workaround would still allow moving the function around as a single piece, but it would not allow modifications within the function itself, which is what we want. The nops elimination pass was mentioned in particular as a potential corner case (there wasn't 100% certainty).

Testing with BOLT shows that after all analyses are complete, when we are ready to `runAllPasses`, any relevant `BinaryFunction` would return false for its `hasCFG()` member function (reflecting just it's `BinaryFunction::State`).
And such `BinaryFunction` would have an empty list of basic blocks, meaning `RemoveNops` or any other intra-procedural pass wouldn't perform any changes that could potentially affect jump table offsets.

So it looks be safe.

---

Other than my above discussion on tests, it LGTM.

kaadam commented:
Thanks the review for you all, I create an minimized assembly test. I cannot argue with that it deeply depend on how compiler generates the code.

kaadam committed:
Added assembly test

kaadam committed:
Fix the comment

kaadam commented:
> Hi! I don't really mind to add stubs like this, although I would prefer asm test, since it is too deeply relies on the compiler code generation. Let also @rafaelauler tell his opinion on this :)

Hi @yota9, I've updated the PR. I enclosed an asm test instead of C one. Could you take a look at it again, please? Thanks Adam.

paschalis-mpeis commented on bolt/test/AArch64/test-indirect-branch.s:
```diff
@@ -0,0 +1,107 @@
+// AArch64MCPlus builder.
+// This test checks that case when we have no shift amount after add
+// instruction. This pattern comes from libc, so needs to build '-static'
+// binary to reproduce the issue easily.
```
nit: maybe adjust a bit these comments from the previous c test.

kaadam committed:
Added a detailed description into the test

yota9 commented:
LGTM, but maybe handle nop+adr case in this patch? It shouldn't be difficult

kaadam commented:
> LGTM, but maybe handle nop+adr case in this patch? It shouldn't be difficult

Hi @yota9, Thanks for your comment. So You mean that to also detect 'nop' after adr instruction in 'analyzeIndirectBranchFragment' following this pattern below? Or are you thinking of a completely different pattern? 
```
  nop   => nop/adr instead of adrp/add
  adr     x13, 0x215a18 <_nl_value_type_LC_COLLATE+0x50>
  ldrh    w13, [x13, w12, uxtw #1]
  adr     x12, 0x247b30 <__gettextparse+0x5b0>
  add     x13, x12, w13, sxth #2
  br      x13 ```

paschalis-mpeis commented on bolt/test/AArch64/test-indirect-branch.s:
```diff
@@ -0,0 +1,110 @@
+// Test how BOLT handles indirect branch sequence of instructions in
+// AArch64MCPlus builder.
+// This test checks the pattern where there is no shift amount after add
```
nit: replace:
```cpp
// This test checks the pattern where there is no shift amount after add
// instruction. The pattern come from libc, it can be reproduced with
// a 'static' built binary.
```

with something like (first line is an empty comment to separate from above):
```cpp
//
// Pattern 1: there is no shift amount after the 'add' instruction.

```

kaadam replied:
Done.

paschalis-mpeis commented on bolt/test/AArch64/test-indirect-branch.s:
```diff
@@ -0,0 +1,110 @@
+//   add     x6, x6, w7, sxtw => no shift amount
+//   br      x6
+//
+// It also tests another case where there is no adrp/add pair.
```
nit: replace
```cpp
// It also tests another case where there is no adrp/add pair.
// The pattern also come from libc, and it only represents in the binary
// if the lld linker is used to create the static binary.
// It doesn't occur with GCC ld linker.
```

with something like:
```cpp
// Pattern 2: nop/adr pair is used in place of adrp/add
```

kaadam replied:
Done.

paschalis-mpeis replied:
Hey @kaadam ,
thanks for the changes, looks good.
Applying the latest suggestion from @yota9 and then we are good to go!

kaadam committed:
Minimize the test case for better comprehension

kaadam commented:
@yota9 @paschalis-mpeis I've minimized the test case (just reorganized and removed superfluous instructions). Updated the PR (commit) description as well. What do you think?

yota9 commented:
@kaadam Hi, could you please reformat test to smth like:

```
  .section .text
  .align 4
  .globl _start
  .type  _start, %function
_start:
  bl test1
  bl test2

 .globl test1
 .type test2, %function
test1:
  adr     x3, jump_table
  ldrh    w3, [x3, x1, lsl #1]
  adr     x1, .test1_0
  add     x3, x1, w3, sxth #2
  br      x3

test1_0:
  ret
test1_1:
  ret

 .globl test2
 .type test2, %function
test2:
  nop
  adr     x3, jump_table
  ldrh    w3, [x3, x1, lsl #1]
  adr     x1, .case0
  add     x3, x1, w3, sxth #2

  .section .rodata,"a",@progbits
jump_table:
  .hword  (.case0-.case0)>>2
  .hword  (.case1-.case0)>>2

test2_0:
  ret
rest2_1:
  ret
test2_2:
  ret

datatable:
  .word test2_0-datatable
  .word test2_1-datatable
  .word test2_2-datatable
```

So they won't be connected, this way it would be much more readable

kaadam commented:
@yota9 @paschalis-mpeis Thanks for your review. I've updated the test based on yota suggestion. I had to pass  '--strict' option to to llvm-bolt. (Since I got 'BOLT-ERROR: Cannot relax adr in non-simple function test1/test2', and test failed). Could you take a look at it again, please?

yota9 reviewed (approved):
Hi! Thanks for test changing. I'm approving this tmp patch since handling the functions with IndirectBranchType::UNKNOWN is still better than aborting on these cases.

kaadam commented:
> Hi! Thanks for test changing. I'm approving this tmp patch since handling the functions with IndirectBranchType::UNKNOWN is still better than aborting on these cases.

Thanks @yota9!
I'm not allowed to submit the PR, @yota9, @paschalis-mpeis may I ask one of you to merge the PR, if the PR has met the criteria for merging?
